iceoryx_hoofs/concurrent/buffer/include/iox/detail/spsc_sofi.inl

```diff
@@ -65,122 +88,132 @@ inline bool SpscSofi<ValueType, CapacityValue>::setCapacity(const uint64_t newSi
+    // - operation reordering:
+    //    - cannot move below if statement otherwise, the code won't compile
+    //    - if it moves above, we might get an outdated read position that will be caught by the
+    //    compare_exchange check
+    uint64_t currentReadPos = m_readPosition.load(std::memory_order_relaxed);
+
+    // Check if queue is full: since we have an extra element (INTERNAL_CAPACITY_ADD_ON), we need to
+    // check if there is a free position for the *next* write position
+    if (nextWritePos < currentReadPos + m_size)
     {
         return !SOFI_OVERFLOW;
     }
 
-    // this is an overflow situation, which means that the next push has no free position, therefore the oldest value
-    // needs to be passed back to the caller
-
-    uint64_t nextReadPosition = currentReadPosition + 1U;
-
-    // we need to update the read position
-    // a) it works, then we need to pass the overflow value back
-    // b) it doesn't work, which means that the pop thread already took the value in the meantime an no further action
-    // is required
-    // memory order success is memory_order_acq_rel
-    //   - this is to prevent the reordering of m_writePosition.store(...) after the increment of the m_readPosition
-    //     - in case of an overflow, this might result in the pop thread getting one element less than the capacity of
-    //       the SoFi if the push thread is suspended in between this two statements
-    //     - it's still possible to get more elements than the capacity, but this is an inherent issue with concurrent
-    //       queues and cannot be prevented since there can always be a push during a pop operation
-    //   - another issue might be that two consecutive pushes (not concurrent) happen on different CPU cores without
-    //     synchronization, then the memory also needs to be synchronized for the overflow case
-    // memory order failure is memory_order_relaxed since there is no further synchronization needed if there is no
-    // overflow
+    // This is an overflow situation so we will need to read the overwritten value
+    // however, it could be that pop() was called in the meantime, i.e. m_readPosition was increased.
+    // Memory order relaxed is enough for both success and failure cases since:
+    //  - synchronization is not needed with m_readPosition
+    //  - operation reordering cannot happen
+    // ======================================
+    // ABA problem: m_readPosition is an uint64_t. Assuming a thread is popping at a rate of 1 GHz while
+    // this thread is blocked, we would still need more than 500 years to overflow m_readPosition and
+    // encounter the ABA problem
     if (m_readPosition.compare_exchange_strong(
-            currentReadPosition, nextReadPosition, std::memory_order_acq_rel, std::memory_order_relaxed))
+            currentReadPos, currentReadPos + 1U, std::memory_order_relaxed, std::memory_order_relaxed))
```

Comment:
Please revert all the changes from this method. I would bet the crate of beer @budrus still owes me that `test_stress_spsc_sofi.cpp` will fail when run on a CPU with a weakly ordered CPU like on an armv8.

Keep in mind, the CPU is allowed to do quite a lot of reordering, e.g.
```cpp
uint64_t currentWritePos = m_writePosition.load(std::memory_order_acquire);
uint64_t nextWritePos = currentWritePos + 1U;

uint64_t currentReadPos = m_readPosition.load(std::memory_order_relaxed);
if (nextWritePos < currentReadPos + m_size) {} // woohoo, there is an overflow and I do not need execute 'm_data[currentWritePos % m_size] = valueIn;' and 'm_writePosition.store(nextWritePos, std::memory_order_release);' and return; lets continue the ride

if (m_readPosition.compare_exchange_strong(
            currentReadPos, currentReadPos + 1U, std::memory_order_relaxed, std::memory_order_relaxed))
{
// hehe, take this INTERNAL_CAPACITY_ADDON, I'm still able to break the contract to have at least CapacityValue of data in a full queue ;)

std::memcpy(&valueOut, &m_data[currentReadPos % m_size], sizeof(ValueType));

// oh no, I have to return; guess it is time to write the value
m_data[currentWritePos % m_size] = valueIn;
m_writePosition.store(nextWritePos, std::memory_order_release);

return SOFI_OVERFLOW;
```